### PR TITLE
First step of optimizing PodAffinity priority function

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #26144 

This is obviously only a first step - I will continue working on this code. However, this is changing the general scheme of computations to what is described in: https://github.com/kubernetes/kubernetes/issues/26144#issuecomment-232612384

### CODE REVIEW

File: plugin/pkg/scheduler/algorithm/priorities/util/topologies.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/*
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/util/topologies.go]: @davidopp - I splitted non_zero.go file into two parts - this is mostly copy of that with some small changes.

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 2]
```diff
@@ -98,99 +87,85 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nod
...
-		totalCount := 0
-		// count weights for the weighted pod affinity
-		if affinity.PodAffinity != nil {
-			for _, weightedTerm := range affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
-				weightedCount, err := ipa.CountWeightByPodMatchAffinityTerm(pod, allPods, weightedTerm.Weight, weightedTerm.PodAffinityTerm, node)
-				if err != nil {
-					return nil, err
+	var maxCount float64
+	var minCount float64
+	counts := make(map[string]float64, len(nodes))
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: write a comment explaining what the map key and value represent (I realize it's the same thing as previously)

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 3]
```diff
@@ -98,99 +87,85 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nod
...
+			return nil, err
+		}
+		existingPodAffinity, err := api.GetAffinityFromPodAnnotations(existingPod.Annotations)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check whether <existingPod> matches <pod> affinity/antiaffinity terms.
+		if affinity.PodAffinity != nil {
+			terms := affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: Here I'd make a more detailed comment, like
For every soft pod affinity term of \<pod\>, if \<existingPod\> matches the term, increment \<counts\> for every node in the cluster with the same term.TopologyKey value as that of \<existingPod\s>'s node by the term's weight.
(I know comments aren't supposed to say "how" but the code is sufficiently complicated that I think this merits an exception.)

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 4]
```diff
@@ -98,99 +87,85 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nod
...
-					if err != nil {
-						return nil, err
-					}
-					if match {
-						totalCount += epWeightedTerm.Weight
-					}
-				}
+		}
+		if affinity.PodAntiAffinity != nil {
+			terms := affinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: Similar comment here, like:
For every soft pod anti-affinity term of \<pod\>, if \<existingPod\> matches the term, decrement \<counts\> for every node in the cluster with the same term.TopologyKey value as that of \<existingPod\s>'s node by the term's weight.

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 5]
```diff
@@ -98,99 +87,85 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nod
...
-				for _, epWeightedTerm := range epAffinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
-					match, err := ipa.failureDomains.CheckIfPodMatchPodAffinityTerm(pod, ep, epWeightedTerm.PodAffinityTerm,
-						func(pod *api.Pod) (*api.Node, error) { return node, nil },
-						func(ep *api.Pod) (*api.Node, error) { return ipa.info.GetNodeInfo(ep.Spec.NodeName) },
-					)
-					if err != nil {
+		// Check whether <pod> matches <existingPod> affinity/antiaffinity terms.
+		if existingPodAffinity.PodAffinity != nil {
+			if ipa.hardPodAffinityWeight > 0 {
+				terms := existingPodAffinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: For every hard pod affinity term of \<existingPod\>, if \<pod\> matches the term, increment \<counts\> for every node in the cluster with the same term.TopologyKey value as that of \<existingPod\s>'s node by the constant \<ipa.hardPodAffinityWeight\>

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 6]
```diff
@@ -98,99 +87,85 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nod
...
+				for _, term := range terms {
+					if err := processTerm(&term, existingPod, pod, existingPodNode, float64(ipa.hardPodAffinityWeight)); err != nil {
 						return nil, err
 					}
-					if match {
-						totalCount -= epWeightedTerm.Weight
-					}
 				}
 			}
+			terms := existingPodAffinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: For every soft pod affinity term of \<existingPod\>, if \<pod\> matches the term, increment \<counts\> for every node in the cluster with the same TopologyKey value as that of \<existingPod\>'s node by the term's weight.

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 7]
```diff
@@ -98,99 +87,85 @@ func (ipa *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nod
...
+			if err := processTerms(terms, existingPod, pod, existingPodNode, 1); err != nil {
+				return nil, err
+			}
 		}
+		if existingPodAffinity.PodAntiAffinity != nil {
+			terms := existingPodAffinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: For every soft pod anti-affinity term of \<existingPod\>, if \<pod\> matches the term, decrement \<counts\> for every node in the cluster with the same term.TopologyKey value as that of \<existingPod\>'s node by the term's weight.

**Comment by davidopp** [comment]
My only comments are suggestions for additional comments to add, so that if the next person who reads this code is as dumb as me, it will at least take them less time to understand it than it did for me. :)

After you add the comments, you can apply LGTM label.

Thanks!

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 7 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 6 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 5 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 4 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 3 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 2 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

**Comment by wojtek-t** [comment]
Comments applied - thanks @davidopp !

**Pull request merged**